File: ruby/translator.rb
require_relative 'braille_alphabet'

input = ARGV.join(" ")

# detects lanugage
def is_braille?(input)
  return false unless input.length % 6 == 0 # if the input is not divisible by 6, it's not braille
  input.chars.each_slice(6).all? { |chunk| BRAILLE_ALPHABET.value?(chunk.join) }
end

# translates braille to english
def translate_braille_to_eng(input)
  translated_string = ""
  capitalize = false
  numberize = false
  decimalize = false

  input.chars.each_slice(6).each do |chunk|
    symbol = chunk.join

    if capitalize
      translated_string << BRAILLE_ALPHABET.key(symbol).upcase
      capitalize = false
    elsif symbol == ".....O"
      capitalize = true
    elsif numberize && symbol == "......"
      translated_string << BRAILLE_ALPHABET.key(symbol)
      numberize = false
    elsif decimalize
      translated_string << BRAILLE_ALPHABET.key(symbol)
      decimalize = false
    elsif symbol == ".O...O"
      decimalize = true
    elsif numberize
      translated_string << BRAILLE_NUMBERS.key(symbol)
    elsif symbol == ".O.OOO"
      numberize = true
    else
      translated_string << BRAILLE_ALPHABET.key(symbol)
    end
  end
  translated_string
end

# translates english to braille
def translate_eng_to_braille(input)
  translated_string = ""

  input.chars.each do |char|
    if BRAILLE_NUMBERS.key?(char)
      translated_string << BRAILLE_ALPHABET["num"]
      translated_string << BRAILLE_NUMBERS[char]
    elsif char == char.upcase && BRAILLE_ALPHABET.key?(char.downcase)
      translated_string << BRAILLE_ALPHABET["cap"]
      translated_string << BRAILLE_ALPHABET[char.downcase]
    elsif BRAILLE_ALPHABET.key?(char)
      translated_string << BRAILLE_ALPHABET[char]
    else
      return "Invalid Input. Please enter a valid English string."
    end
  end
  translated_string
end

# This app doesn't account for foreign languages, gibberish phrases or phrases consisting of only special characters.
# It's possible a user might want to check specific letters, or characters so strings like "<<>//..<" are considered english.
# To ensure english only input we could check against an english dictionary, or use a gem like 'lingua' to detect the language.
# This check will reject any input including a character not in the BRAILLE_ALPHABET or BRAILLE_NUMBERS hashes.

if is_braille?(input)
puts translate_braille_to_eng(input)
else
puts translate_eng_to_braille(input)
end

File: javascript/translator.js
const brailleMap = {
    'A': 'O.....', 'B': 'O.O...', 'C': 'OO....', 'D': 'OO.O..', 'E': 'O..O..',
    'F': 'OOO...', 'G': 'OOOO..', 'H': 'O.OO..', 'I': '.OO...', 'J': '.OOO..',
    'K': 'O...O.', 'L': 'O.O.O.', 'M': 'OO..O.', 'N': 'OO.OO.', 'O': 'O..OO.',
    'P': 'OOO.O.', 'Q': 'OOOOO.', 'R': 'O.OOO.', 'S': '.OO..O', 'T': '.OOO.O',
    'U': 'O...OO', 'V': 'O.O.OO', 'W': '.OOO.O', 'X': 'OO..OO', 'Y': 'OO.OOO',
    'Z': 'O..OOO', ' ': '......',
    '1': '.O.....', '2': '.O.O...', '3': '.OO....', '4': '.OO.O..', '5': '.O..O..',
    '6': '.OOO...', '7': '.OOOO..', '8': '.O.OO..', '9': '..OO...', '0': '..OOO..',
};

function translate(input) {
    let output = '';
    const isBraille = input.split('').every(char => char === 'O' || char === '.');

    if (isBraille) {
        // Translate from Braille to English
        const brailleChars = input.match(/.{1,6}/g);
        for (const brailleChar of brailleChars) {
            const englishChar = Object.keys(brailleMap).find(key => brailleMap[key] === brailleChar);
            output += englishChar ? englishChar : '?'; // Handle unknown Braille characters
        }
    } else {
        // Translate from English to Braille
        const upperInput = input.toUpperCase();
        for (const char of upperInput) {
            if (brailleMap[char]) {
                output += brailleMap[char]; // Add corresponding Braille
            } else {
                output += '......'; // For any unknown characters
            }
        }
    }
    return output;
}

// Get input from command line
const input = process.argv.slice(2).join(' ') || '';
console.log(translate(input));